File: harness/benchmarking/python/quicksort.py
#!/usr/bin/env python3
"""
QuickSort Benchmark - Python Implementation
O(n log n) average case performance validation
"""

import sys
import time
import json
import random
from typing import List, Any

def quicksort(arr: List[int]) -> List[int]:
    """Standard quicksort implementation"""
    if len(arr) <= 1:
        return arr
    
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    
    return quicksort(left) + middle + quicksort(right)

def quicksort_inplace(arr: List[int], low: int = 0, high: int = None) -> None:
    """In-place quicksort implementation"""
    if high is None:
        high = len(arr) - 1
    
    if low < high:
        pi = partition(arr, low, high)
        quicksort_inplace(arr, low, pi - 1)
        quicksort_inplace(arr, pi + 1, high)

def partition(arr: List[int], low: int, high: int) -> int:
    """Partition helper for in-place quicksort"""
    pivot = arr[high]
    i = low - 1
    
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def generate_test_array(size: int, pattern: str) -> List[int]:
    """Generate test arrays with different patterns"""
    if pattern == "random":
        arr = list(range(size))
        random.shuffle(arr)
        return arr
    elif pattern == "sorted":
        return list(range(size))
    elif pattern == "reverse":
        return list(range(size - 1, -1, -1))
    elif pattern == "partial":
        arr = list(range(size))
        # Shuffle last quarter
        quarter = size // 4
        tail = arr[-quarter:]
        random.shuffle(tail)
        arr[-quarter:] = tail
        return arr
    else:
        return [0] * size

def benchmark_quicksort(iterations: int, size: int, pattern: str, inplace: bool = True) -> float:
    """Benchmark quicksort with specified parameters"""
    total_time = 0.0
    
    for _ in range(iterations):
        arr = generate_test_array(size, pattern)
        
        start_time = time.perf_counter()
        if inplace:
            quicksort_inplace(arr.copy())
        else:
            sorted_arr = quicksort(arr)
        elapsed = time.perf_counter() - start_time
        
        total_time += elapsed
    
    return (total_time / iterations) * 1_000_000  # Convert to microseconds

def main():
    # Parse command line arguments
    iterations = 1000
    if len(sys.argv) > 2 and sys.argv[1] == "--iterations":
        try:
            iterations = int(sys.argv[2])
        except ValueError:
            pass
    
    results = {
        "algorithm": "quicksort",
        "language": "python",
        "iterations": iterations,
        "results": []
    }
    
    test_sizes = [10, 50, 100, 500, 1000, 5000]
    patterns = ["random", "sorted", "reverse", "partial"]
    
    for size in test_sizes:
        for pattern in patterns:
            avg_time = benchmark_quicksort(iterations, size, pattern)
            
            results["results"].append({
                "size": size,
                "pattern": pattern,
                "avg_time_us": round(avg_time, 3)
            })
    
    print(json.dumps(results, indent=2))

if __name__ == "__main__":
    main()

File: harness/benchmarking/javascript/quicksort.js
#!/usr/bin/env node
/**
 * QuickSort Benchmark - JavaScript Implementation
 * O(n log n) average case performance validation
 */

function quicksort(arr) {
    if (arr.length <= 1) return arr;
    
    const pivot = arr[Math.floor(arr.length / 2)];
    const left = arr.filter(x => x < pivot);
    const middle = arr.filter(x => x === pivot);
    const right = arr.filter(x => x > pivot);
    
    return [...quicksort(left), ...middle, ...quicksort(right)];
}

function quicksortInplace(arr, low = 0, high = null) {
    if (high === null) high = arr.length - 1;
    
    if (low < high) {
        const pi = partition(arr, low, high);
        quicksortInplace(arr, low, pi - 1);
        quicksortInplace(arr, pi + 1, high);
    }
    
    return arr;
}

function partition(arr, low, high) {
    const pivot = arr[high];
    let i = low - 1;
    
    for (let j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
    }
    
    [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
    return i + 1;
}

function generateTestArray(size, pattern) {
    switch (pattern) {
        case 'random': {
            const arr = Array.from({length: size}, (_, i) => i);
            // Fisher-Yates shuffle
            for (let i = arr.length - 1; i > 0; i--) {
                const j = Math.floor(Math.random() * (i + 1));
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
            return arr;
        }
        case 'sorted':
            return Array.from({length: size}, (_, i) => i);
        case 'reverse':
            return Array.from({length: size}, (_, i) => size - 1 - i);
        case 'partial': {
            const arr = Array.from({length: size}, (_, i) => i);
            // Shuffle last quarter
            const quarter = Math.floor(size / 4);
            const start = size - quarter;
            for (let i = size - 1; i >= start; i--) {
                const j = Math.floor(Math.random() * (i - start + 1)) + start;
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
            return arr;
        }
        default:
            return new Array(size).fill(0);
    }
}

function benchmarkQuicksort(iterations, size, pattern, inplace = true) {
    let totalTime = 0;
    
    for (let i = 0; i < iterations; i++) {
        const arr = generateTestArray(size, pattern);
        
        const startTime = process.hrtime.bigint();
        if (inplace) {
            quicksortInplace([...arr]); // Copy array to preserve original
        } else {
            quicksort(arr);
        }
        const endTime = process.hrtime.bigint();
        
        totalTime += Number(endTime - startTime);
    }
    
    return totalTime / iterations / 1000; // Convert to microseconds
}

function main() {
    // Parse command line arguments
    let iterations = 1000;
    const args = process.argv.slice(2);
    
    if (args.length >= 2 && args[0] === '--iterations') {
        iterations = parseInt(args[1]) || 1000;
    }
    
    const results = {
        algorithm: 'quicksort',
        language: 'javascript',
        iterations: iterations,
        results: []
    };
    
    const testSizes = [10, 50, 100, 500, 1000, 5000];
    const patterns = ['random', 'sorted', 'reverse', 'partial'];
    
    for (const size of testSizes) {
        for (const pattern of patterns) {
            const avgTime = benchmarkQuicksort(iterations, size, pattern);
            
            results.results.push({
                size: size,
                pattern: pattern,
                avg_time_us: Math.round(avgTime * 1000) / 1000
            });
        }
    }
    
    console.log(JSON.stringify(results, null, 2));
}

if (require.main === module) {
    main();
}